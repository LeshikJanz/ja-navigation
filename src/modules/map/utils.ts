export function init() {
  window.jaMap = new window.ymaps.Map("map", {
    center: [55.76, 37.64],
    zoom: 7
  });
}

export const runMap = () =>
  window.ymaps.ready(init)
