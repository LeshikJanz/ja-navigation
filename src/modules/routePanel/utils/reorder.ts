import { IWaypoint } from "../types";

export const reorder = (list: IWaypoint[], startIndex: number, endIndex: number): IWaypoint[] => {
  const result = Array.from(list);
  const [removed] = result.splice(startIndex, 1);
  result.splice(endIndex, 0, removed);

  return result;
};
