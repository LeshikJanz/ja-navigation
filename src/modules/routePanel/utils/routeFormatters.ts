import { IWaypoint } from "../types";
import { IGeoObject, GeoObject } from "yandex-maps";

const minPointsCount = 2

export const getPlacemarkId = (placemark: IGeoObject | any) => placemark.properties &&
  placemark.properties._data &&
  placemark.properties._data.iden

const makeRouteInstance = () => {
  let placemarks: IGeoObject[] = []
  let route: IGeoObject

  const addPlacemark = (waypoint: IWaypoint) => {
    const placemark = new window.ymaps.Placemark(waypoint.coords, {
      balloonContent: waypoint.value,
      iden: waypoint.id,
    }, {
        draggable: true
      });

    placemarks.push(placemark);
    window.jaMap.geoObjects.add(placemark);
    return placemark
  };

  const removePlacemark = (waypointId: string) => {
    const placemark = placemarks.find(placemark => getPlacemarkId(placemark) === waypointId)
    placemarks = placemarks.filter(p => p !== placemark)
    window.jaMap.geoObjects.remove(placemark)
  };

  const assignDragEndEvent = async (placemark: GeoObject, updateWaypoint: any) => {
    placemark.events.add("dragend", async (e: any) => {
      const thisPlacemark = e.get("target");
      const coords = thisPlacemark.geometry.getCoordinates();
      const { address } = await geocode(coords)
      const waypoints = updateWaypoint(getPlacemarkId(placemark), coords, address)
      routeInstance.updateRoute(waypoints, updateWaypoint)
    });
  };

  const geocode = (request: string) => {
    try {
      return window.ymaps.geocode(request, { results: 1 }).then((res: any) => {
        if (!res.geoObjects) throw new Error("Bad address");
        const geoObject = res.geoObjects.get(0);
        if (!geoObject) throw new Error("Bad address");
        const address = geoObject.getAddressLine();
        const coords = geoObject.geometry.getCoordinates();
        const bounds = geoObject.properties.get("boundedBy");
        return { address, coords, bounds };
      });
    } catch (errors) {
      console.error("errors", errors)
    }
  };

  const createRoute = (waypoints: IWaypoint[]) => {
    if (waypoints.length < minPointsCount) return
    route = new window.ymaps.multiRouter.MultiRoute(
      {
        referencePoints: waypoints.map(waypoint => waypoint.coords),
        params: { results: minPointsCount }
      },
      { boundsAutoApply: true }
    );
    window.jaMap.geoObjects.add(route);
  }

  const updateRoute = (waypoints: IWaypoint[], updateWaypoint: any) => {
    window.jaMap.geoObjects.removeAll()
    placemarks = []
    waypoints.forEach(waypoint => assignDragEndEvent(addPlacemark(waypoint), updateWaypoint))
    createRoute(waypoints)
  }

  return { addPlacemark, removePlacemark, geocode, assignDragEndEvent, createRoute, updateRoute, placemarks }
};

export const routeInstance = makeRouteInstance()
