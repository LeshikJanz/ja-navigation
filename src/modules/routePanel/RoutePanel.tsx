import React, { FormEvent, Fragment } from "react";
import WaypointList from "./WaypointList";
import { reorder } from "./utils";

import { IWaypoint } from "./types";
import { DropResult } from "react-beautiful-dnd";
import SearchBox from "./SearchBox";

class RoutePanel extends React.Component<
  { isMapInit: boolean },
  { waypoints: IWaypoint[] }
> {
  state = {
    waypoints: []
  };

  addDestination = (destination: string, coords: [string, string]): void => {
    console.log("coords", coords);
    this.setState(prevState => ({
      waypoints: [
        ...prevState.waypoints,
        {
          id: `${new Date().getTime()}`,
          value: destination,
          coords
        }
      ]
    }));
  };

  deleteWaypoint = (id: string) =>
    this.setState(prevState => ({
      waypoints: prevState.waypoints.filter(waypoint => waypoint.id !== id)
    }));

  reorderWaypoints = (result: DropResult) => {
    if (!result.destination) return;

    const waypoints = reorder(
      this.state.waypoints,
      result.source.index,
      result.destination.index
    );

    this.setState({ waypoints });
  };

  render() {
    if (!this.props.isMapInit) return null;
    const { waypoints } = this.state;
    return (
      <Fragment>
        <SearchBox waypoints={waypoints} addDestination={this.addDestination} />
        <WaypointList
          waypoints={waypoints}
          reorderWaypoints={this.reorderWaypoints}
          deleteWaypoint={this.deleteWaypoint}
        />
      </Fragment>
    );
  }
}

export default RoutePanel;
