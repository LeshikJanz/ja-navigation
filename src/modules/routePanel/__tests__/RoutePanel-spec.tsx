import React from "react";
import { shallow } from "enzyme";
import RoutePanel from "../RoutePanel";
import { routeInstance } from "../utils/routeFormatters";

describe("RoutePanel component", () => {
  it("matches snapshot with isMapInit = false", () => {
    const props = {
      isMapInit: false
    };
    const component = shallow(<RoutePanel {...props} />);
    expect(component).toMatchSnapshot();
  });

  it("matches snapshot with isMapInit = true", () => {
    const props = {
      isMapInit: true
    };
    const component = shallow(<RoutePanel {...props} />);
    expect(component).toMatchSnapshot();
  });

  it("should addDestination add waypoint to state", () => {
    const props = {
      isMapInit: true
    };
    const component = shallow(<RoutePanel {...props} />);
    expect(component.state().waypoints.length).toBe(0);
    component.instance().addDestination("Minsk", ["52.52", "22.55"], "");
    expect(component.state().waypoints.length).toBe(1);
  });

  it("should deleteWaypoint call routeInstance.removePlacemark", () => {
    const props = {
      isMapInit: true
    };
    routeInstance.removePlacemark = jest.fn();
    const component = shallow(<RoutePanel {...props} />);
    expect(routeInstance.removePlacemark.mock.calls.length).toBe(0);
    component.instance().deleteWaypoint("123");
    expect(routeInstance.removePlacemark.mock.calls.length).toBe(1);
  });

  it("should deleteWaypoint delete waypoint by id", () => {
    const props = {
      isMapInit: true
    };
    const component = shallow(<RoutePanel {...props} />);
    component.setState({
      waypoints: [{ id: "123", value: "Minsk", coords: ["52.34", "55.33"] }]
    });
    expect(component.state().waypoints.length).toBe(1);
    component.instance().deleteWaypoint("123");
    expect(component.state().waypoints.length).toBe(0);
  });

  it("should updateWaypoint update waypoint by id", () => {
    const props = {
      isMapInit: true
    };
    const prevCoords = ["25.25", "52.52"];
    const nextCoords = ["55.55", "22.22"];
    const component = shallow(<RoutePanel {...props} />);
    component.setState({
      waypoints: [{ id: "123", value: "Minsk", coords: prevCoords }]
    });
    expect(component.state().waypoints[0].coords).toBe(prevCoords);
    expect(component.state().waypoints[0].value).toBe("Minsk");
    component.instance().updateWaypoint("123", nextCoords, "Moscow");
    expect(component.state().waypoints[0].coords).toBe(nextCoords);
    expect(component.state().waypoints[0].value).toBe("Moscow");
  });
});
