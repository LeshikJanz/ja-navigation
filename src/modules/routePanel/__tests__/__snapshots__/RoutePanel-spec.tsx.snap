// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoutePanel component matches snapshot with isMapInit = false 1`] = `ShallowWrapper {}`;

exports[`RoutePanel component matches snapshot with isMapInit = true 1`] = `ShallowWrapper {}`;
